test/framework/src/main/java/org/elasticsearch/test/InternalTestCluster.java

```diff
@@ -593,7 +594,8 @@ private NodeAndClient buildNode(int nodeId, long seed, Settings settings,
             .put("node.name", name)
             .put(NodeEnvironment.NODE_ID_SEED_SETTING.getKey(), seed);
 
-        final boolean usingSingleNodeDiscovery = DiscoveryModule.DISCOVERY_TYPE_SETTING.get(finalSettings.build()).equals("single-node");
+        final boolean usingSingleNodeDiscovery = DISCOVERY_TYPE_SETTING.exists(settings) &&
```

Comment:
The issue here is that we are not accessing the true "final" settings of the node. These are just the settings explicitly passed to the node. If the Zen-Disco plugin is installed, it will extend these settings by setting the `discovery.type` to `test-zen`. The default discovery implementation is "single-node" now, which means that if the settings we are using here do not have an explicit ``discovery.type` setting, then the code here thinks that we will be using `single-node`, which is wrong.
It's a bit of a chicken-egg problem. We can only know about the actual setting of "discovery.type" after we've constructed the mock-node, but here we also want to influence the settings of the mock node before constructing it. In theory we could construct a temporary MockNode but that is also very heavy-weight.